packages/components/src/utils/hds-aria-described-by.ts

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import { tracked } from '@glimmer/tracking';
+import { scheduleOnce } from '@ember/runloop';
+import type Component from '@glimmer/component';
+
+type ElementIDMap = Map<HTMLElement, string>;
+
+class AriaDescribedByAttributes {
+  private _attributes = new WeakMap<AriaDescribedByComponent, ElementIDMap>();
+
+  findOrCreateAttributes(obj: AriaDescribedByComponent): ElementIDMap {
+    if (this._attributes.has(obj)) {
+      return this._attributes.get(obj) as ElementIDMap;
+    }
+
+    const attributes: ElementIDMap = new Map();
+    this._attributes.set(obj, attributes);
+    return attributes;
+  }
+
+  register(obj: AriaDescribedByComponent, element: HTMLElement) {
+    const attrs = this.findOrCreateAttributes(obj);
+    attrs.set(element, element.id);
+  }
+
+  unregister(obj: AriaDescribedByComponent, element: HTMLElement) {
+    const attrs = this.findOrCreateAttributes(obj);
+    attrs.delete(element);
+  }
+
+  entries(obj: AriaDescribedByComponent) {
+    const attrs = this.findOrCreateAttributes(obj);
+    return Array.from(attrs.values());
+  }
+}
+
+const ariaDescribedByAttributes = new AriaDescribedByAttributes();
+
+interface AriaDescribedByArgs {
+  Args: {
+    extraAriaDescribedBy: string;
+  };
+}
+
+interface AriaDescribedByComponent extends Component<AriaDescribedByArgs> {
+  __ARIA_DESCRIPTION_IDS__: string[];
+  ariaDescribedBy: string;
+}
+
+// essentially a type that says we return a subclass of the given type T
+type ClassOf<T> = new () => T;
+
+export function ariaDescribedBy(
+  klass: ClassOf<AriaDescribedByComponent>
+): ClassOf<AriaDescribedByComponent> {
+  class X extends klass {
```

Comment:
have to write this as `class X extends klass {}; return X` instead of `return class extends klass {}`. Otherwise typescript complains about using decorators for some reason 🤷